File: src/sys/dotnet/fan/sys/Depend.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   14 Jan 06  Andy Frank  Creation
//

using System.Collections;
using System.Text;

namespace Fan.Sys
{
  /// <summary>
  /// Depend.
  /// </summary>
  public sealed class Depend : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static Depend fromStr(string str) { return fromStr(str, true); }
    public static Depend fromStr(string str, bool check)
    {
      try
      {
        return new Parser(str).parse();
      }
      catch (System.Exception)
      {
        if (!check) return null;
        throw ParseErr.make("Depend", str).val;
      }
    }

    private Depend(string name, Constraint[] constraints)
    {
      this.m_name = name;
      this.m_constraints = constraints;
    }

  //////////////////////////////////////////////////////////////////////////
  // Parser
  //////////////////////////////////////////////////////////////////////////

    class Parser
    {
      internal Parser(string str)
      {
        this.m_str = str;
        this.m_len = str.Length;
        consume();
      }

      internal Depend parse()
      {
        m_name = name();
        constraints.Add(constraint());
        while (m_cur == ',')
        {
          consume();
          consumeSpaces();
          constraints.Add(constraint());
        }
        if (m_pos <= m_len) throw new System.Exception();
        return new Depend(m_name, (Constraint[])constraints.ToArray(
          System.Type.GetType("Fan.Sys.Depend+Constraint")));
      }

      private string name()
      {
        StringBuilder s = new StringBuilder();
        while (m_cur != ' ')
        {
          if (m_cur < 0) throw new System.Exception();
          s.Append((char)m_cur);
          consume();
        }
        consumeSpaces();
        if (s.Length == 0) throw new System.Exception();
        return s.ToString();
      }

      private Constraint constraint()
      {
        Constraint c = new Constraint();
        c.version = version();
        consumeSpaces();
        if (m_cur == '+')
        {
          c.isPlus = true;
          consume();
          consumeSpaces();
        }
        else if (m_cur == '-')
        {
          consume();
          consumeSpaces();
          c.endVersion = version();
          consumeSpaces();
        }
        return c;
      }

      private Version version()
      {
        List segs = new List(Sys.IntType, 4);
        int seg = consumeDigit();
        while (true)
        {
          if ('0' <= m_cur && m_cur <= '9')
          {
            seg = seg*10 + consumeDigit();
          }
          else
          {
            segs.add(Long.valueOf(seg));
            seg = 0;
            if (m_cur != '.') break;
            else consume();
          }
        }
        return new Version(segs);
      }

      private int consumeDigit()
      {
        if ('0' <= m_cur && m_cur <= '9')
        {
          int digit = m_cur - '0';
          consume();
          return digit;
        }
        throw new System.Exception();
      }

      private void consumeSpaces()
      {
        while(m_cur == ' ') consume();
      }

      private void consume()
      {
        if (m_pos < m_len)
        {
          m_cur = m_str[m_pos++];
        }
        else
        {
          m_cur = -1;
          m_pos = m_len+1;
        }
      }

      int m_cur;
      int m_pos;
      int m_len;
      string m_str;
      string m_name;
      ArrayList constraints = new ArrayList(4);
    }

  //////////////////////////////////////////////////////////////////////////
  // .NET
  //////////////////////////////////////////////////////////////////////////

    public override int GetHashCode()
    {
      return toStr().GetHashCode();
    }

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override bool Equals(object obj)
    {
      if (obj is Depend)
        return toStr() == toStr(obj);
      else
        return false;
    }

    public override long hash()
    {
      return FanStr.hash(toStr());
    }

    public override Type @typeof()
    {
      return Sys.DependType;
    }

    public override string toStr()
    {
      if (m_str == null)
      {
        StringBuilder s = new StringBuilder();
        s.Append(m_name).Append(' ');
        for (int i=0; i<m_constraints.Length; i++)
        {
          if (i > 0) s.Append(',');
          Constraint c = m_constraints[i];
          s.Append(c.version);
          if (c.isPlus) s.Append('+');
          if (c.endVersion != null) s.Append('-').Append(c.endVersion);
        }
        m_str = s.ToString();
      }
      return m_str;
    }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public string name()
    {
      return m_name;
    }

    public long size()
    {
      return m_constraints.Length;
    }

    public Version version() { return version(0); }
    public Version version(long index)
    {
      return m_constraints[(int)index].version;
    }

    public bool isPlus() { return isPlus(0); }
    public bool isPlus(long index)
    {
      return m_constraints[(int)index].isPlus;
    }

    public bool isRange() { return isRange(0); }
    public bool isRange(long index)
    {
      return m_constraints[(int)index].endVersion != null;
    }

    public Version endVersion() { return endVersion(0); }
    public Version endVersion(long index)
    {
      return m_constraints[(int)index].endVersion;
    }

    public bool match(Version v)
    {
      for (int i=0; i<m_constraints.Length; i++)
      {
        Constraint c = m_constraints[i];
        if (c.isPlus)
        {
          // versionPlus
          if (c.version.compare(v) <= 0)
            return true;
        }
        else if (c.endVersion != null)
        {
          // versionRange
          if (c.version.compare(v) <= 0 &&
              (c.endVersion.compare(v) >= 0 || match(c.endVersion, v)))
            return true;
        }
        else
        {
          // versionSimple
          if (match(c.version, v))
            return true;
        }
      }
      return false;
    }

    private static bool match(Version a, Version b)
    {
      if (a.segments().sz() > b.segments().sz()) return false;
      for (int i=0; i<a.segments().sz(); ++i)
        if (a.segment(i) != b.segment(i))
          return false;
      return true;
    }

  //////////////////////////////////////////////////////////////////////////
  // Constraint
  //////////////////////////////////////////////////////////////////////////

    internal class Constraint
    {
      internal Version version = null;
      internal bool isPlus = false;
      internal Version endVersion = null;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    private readonly string m_name;
    private readonly Constraint[] m_constraints;
    private string m_str;

  }
}

File: src/sys/java/fan/sys/Depend.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   3 Nov 06  Brian Frank  Creation
//
package fan.sys;

import java.util.*;

/**
 * Depend
 */
public final class Depend
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Depend fromStr(String str) { return fromStr(str, true); }
  public static Depend fromStr(String str, boolean checked)
  {
    try
    {
      return new Parser(str).parse();
    }
    catch (Throwable e)
    {
      if (!checked) return null;
      throw ParseErr.make("Depend", str);
    }
  }

  private Depend(String name, Constraint[] constraints)
  {
    this.name = name;
    this.constraints = constraints;
  }

//////////////////////////////////////////////////////////////////////////
// Parser
//////////////////////////////////////////////////////////////////////////

  static class Parser
  {
    Parser(String str)
    {
      this.str = str;
      this.len = str.length();
      consume();
    }

    Depend parse()
    {
      name = name();
      constraints.add(constraint());
      while (cur == ',')
      {
        consume();
        consumeSpaces();
        constraints.add(constraint());
      }
      if (pos <= len) throw new RuntimeException();
      return new Depend(name, (Constraint[])constraints .toArray(new Constraint[constraints.size()]));
    }

    private String name()
    {
      StringBuilder s = new StringBuilder();
      while (cur != ' ')
      {
        if (cur < 0) throw new RuntimeException();
        s.append((char)cur);
        consume();
      }
      consumeSpaces();
      if (s.length() == 0) throw new RuntimeException();
      return s.toString();
    }

    private Constraint constraint()
    {
      Constraint c = new Constraint();
      c.version = version();
      consumeSpaces();
      if (cur == '+')
      {
        c.isPlus = true;
        consume();
        consumeSpaces();
      }
      else if (cur == '-')
      {
        consume();
        consumeSpaces();
        c.endVersion = version();
        consumeSpaces();
      }
      return c;
    }

    private Version version()
    {
      List segs = new List(Sys.IntType, 4);
      int seg = consumeDigit();
      while (true)
      {
        if ('0' <= cur && cur <= '9')
        {
          seg = seg*10 + consumeDigit();
        }
        else
        {
          segs.add(Long.valueOf(seg));
          seg = 0;
          if (cur != '.') break;
          else consume();
        }
      }
      return new Version(segs);
    }

    private int consumeDigit()
    {
      if ('0' <= cur && cur <= '9')
      {
        int digit = cur - '0';
        consume();
        return digit;
      }
      throw new RuntimeException();
    }

    private void consumeSpaces()
    {
      while (cur == ' ') consume();
    }

    private void consume()
    {
      if (pos < len)
      {
        cur = str.charAt(pos++);
      }
      else
      {
        cur = -1;
        pos = len+1;
      }
    }

    int cur;
    int pos;
    int len;
    String str;
    String name;
    ArrayList constraints = new ArrayList(4);
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public boolean equals(Object obj)
  {
    if (obj instanceof Depend)
      return toStr().equals(toStr(obj));
    else
      return false;
  }

  public int hashCode()
  {
    return toStr().hashCode();
  }

  public long hash()
  {
    return FanStr.hash(toStr());
  }

  public Type typeof()
  {
    return Sys.DependType;
  }

  public String toStr()
  {
    if (str == null)
    {
      StringBuilder s = new StringBuilder();
      s.append(name).append(' ');
      for (int i=0; i<constraints.length; ++i)
      {
        if (i > 0) s.append(',');
        Constraint c = constraints[i];
        s.append(c.version);
        if (c.isPlus) s.append('+');
        if (c.endVersion != null) s.append('-').append(c.endVersion);
      }
      str = s.toString();
    }
    return str;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public final String name()
  {
    return name;
  }

  public final long size()
  {
    return constraints.length;
  }

  public final Version version() { return version(0L); }
  public final Version version(long index)
  {
    return constraints[(int)index].version;
  }

  public final boolean isPlus() { return isPlus(0L); }
  public final boolean isPlus(long index)
  {
    return constraints[(int)index].isPlus;
  }

  public final boolean isRange() { return isRange(0L); }
  public final boolean isRange(long index)
  {
    return constraints[(int)index].endVersion != null;
  }

  public final Version endVersion() { return endVersion(0L); }
  public final Version endVersion(long index)
  {
    return constraints[(int)index].endVersion;
  }

  public final boolean match(Version v)
  {
    for (int i=0; i<constraints.length; ++i)
    {
      Constraint c = constraints[i];
      if (c.isPlus)
      {
        // versionPlus
        if (c.version.compare(v) <= 0)
          return true;
      }
      else if (c.endVersion != null)
      {
        // versionRange
        if (c.version.compare(v) <= 0 &&
            (c.endVersion.compare(v) >= 0 || doMatch(c.endVersion, v)))
          return true;
      }
      else
      {
        // versionSimple
        if (doMatch(c.version, v))
          return true;
      }
    }
    return false;
  }

  private static boolean doMatch(Version a, Version b)
  {
    if (a.segments().sz() > b.segments().sz()) return false;
    for (int i=0; i<a.segments().sz(); ++i)
      if (a.segment(i) != b.segment(i))
        return false;
    return true;
  }

//////////////////////////////////////////////////////////////////////////
// Constraint
//////////////////////////////////////////////////////////////////////////

  static class Constraint
  {
    Version version;
    boolean isPlus;
    Version endVersion;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  private final String name;
  private final Constraint[] constraints;
  private String str;

}